go-controller/pkg/crd/adminpolicybasedroute/v1/types.go

```diff
@@ -0,0 +1,125 @@
+type AdminPolicyBasedExternalRouteSpec struct {
+	Policies []*ExternalPolicy `json:"policies"`
+}
+
+type ExternalPolicy struct {
+	From     ExternalNetworkSource `json:"from"`
+	NextHops ExternalNextHops      `json:"nextHops"`
+}
+
+type ExternalNetworkSource struct {
+	NamespaceSelector metav1.LabelSelector `json:"namespaceSelector,omitempty"`
+}
+
+type ExternalNextHops struct {
+	// +kubebuilder:validation:UniqueItems=true
+	StaticHops []*StaticHop `json:"static,omitempty"`
+	// +kubebuilder:validation:UniqueItems=true
+	DynamicHops []*DynamicHop `json:"dynamic,omitempty"`
+}
+
+type StaticHop struct {
+	IP           string `json:"ip,omitempty"`
+	BFDEnabled   bool   `json:"bfdEnabled,omitempty"`
+	SkipHostSNAT bool   `json:"skipHostSNAT,omitempty"`
+}
+
+type DynamicHop struct {
+	PodSelector           metav1.LabelSelector  `json:"podSelector,omitempty"`
+	NamespaceSelector     *metav1.LabelSelector `json:"namespaceSelector,omitempty"`
+	NetworkAttachmentName string                `json:"networkAttachmentName,omitempty"`
+	BFDEnabled            bool                  `json:"bfdEnabled,omitempty"`
+	SkipHostSNAT          bool                  `json:"skipHostSNAT,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// AdminPolicyBasedExternalRouteList contains a list of AdminPolicyBasedExternalRoutes
+type AdminPolicyBasedExternalRouteList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []AdminPolicyBasedExternalRoute `json:"items"`
+}
+
+// AdminPolicyBasedRouteStatus describes the current status of the AdminPolicyBased route types.
+// +kubebuilder:object:root=true
+type AdminPolicyBasedRouteStatus struct {
+	// An array of Human-readable messages indicating details about the status of the object.
+	Messages []string `json:"messages,omitempty"`
+	// A concise indication of whether the AdminPolicyBasedRoute resource is applied or not
+	Status StatusType `json:"status,omitempty"`
+}
+
+type StatusType string
+
+const (
+	SuccessStatus StatusType = "Success"
+	FailStatus    StatusType = "Fail"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:path=adminpolicybasedexternalroute,scope=Cluster
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+type AdminPolicyBasedInternalRoute struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              AdminPolicyBasedInternalRouteSpec `json:"spec,omitempty"`
+	Status            AdminPolicyBasedRouteStatus       `json:"status,omitempty"`
+}
+
+// AdminPolicyBasedInternalRouteSpec defines the desired state of AdminPolicyBasedInternalRoute
+type AdminPolicyBasedInternalRouteSpec struct {
+	Policies []*InternalPolicies `json:"policies"`
+}
+
+type InternalPolicies struct {
+}
```

Comment:
precisely my point, we cannot have half-built unknown APIs like this.

Reply:
Commented out.